Create spacewar.ts

alex-kulcsar opened the pull request:
Submitting a clone of Spacewar! for consideration in the user gallery.

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Microsoft/pxt-arcade?pullRequest=878) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Microsoft/pxt-arcade?pullRequest=878) before we can accept your contribution.<br/><br/>:x: alex-kulcsar [sign now](https://cla.opensource.microsoft.com/Microsoft/pxt-arcade?pullRequest=878)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Microsoft/pxt-arcade?pullRequest=878) it.</sub>

riknoll commented:
We should probably rename it just to be safe... Spacefight? (suggested by @jwunderl)

alex-kulcsar commented:
Fine by me, @jwunderl and @riknoll. I've updated the commit, but the splash screen still includes an appropriate attribution. Let me know if you want me to remove that.

jwunderl commented:
I really like this one!

There's a lot of blocks in there, so I split a few parts up with functions and other minor changes to cut off about 100 blocks to get it down to 675 - still a bit too high to go to typescript and back, but we can work on that a bit. I also accidentally pressed ``Format Code``, so it's laid out a bit differently from the one you linked but should behave the same. https://makecode.com/_A8CA91C4FixV

A couple minor things; 

* it feels like the game tends to end really quickly, as I end up hitting either the star or the other player within about 5-10 seconds most of the time -- maybe we can remove some of the game overs and convert them to just reset the player and give score to someone? Or possibly even just go for a one minute timer, with the goal being to get the most points by then (either by destroying the other player, or reaching the goal a bunch, etc).
* It might be worth just moving the just using ``on game update`` events to handle fuel and rockets instead of being exact about the timing; it cuts off another ~85 blocks, which can be important on computers that aren't high end: https://makecode.com/_PCueTXE5aHMm . The timings a bit off with that (e.g. sometimes you'll fire a rocket and then immediately get one back), but you can regain much of the timing either with some flags (e.g. ignore the first time it would refuel in a row flip the flag, then continue to refuel while the flag is true / set it to false when full) or by increasing the number of fuel and rocket 'units' consumed and obtained (instead of 1 every 5 seconds, use 10 at once but get them every half second, etc)

A couple other tangential things;

* I like your 'draw text' extension a lot, I'll have to take another look at it some time
* For your extensions, you can add a variable dropdown for blocks with ``variables.shadow="variables_get"``, so you don't have holes in the blocks when users drag them out. Related: There's a minor bug preventing them from behaving perfectly - https://github.com/Microsoft/pxt-arcade/issues/409 - but if you omit the initial value in parens they will show up nicely for now
* This game really makes me think we should add a single block to set default values for controller repeat events delay and interval (related: https://github.com/Microsoft/pxt-arcade/issues/767). That way we can cut half the controller events in this game immediately, and also make it easy to use repeat events in a wider variety of contexts

alex-kulcsar commented:
Thanks, Joey (@jwunderl)! Glad you liked! I've added you to the list of credits. :-)

I've done some updates to the gameplay, using your comments as guidance. Give this a try and see what you think as far as gameplay is concerned:

https://makecode.com/_ffxf03UiT2hy

Changes that I've added to your modifications:
- Shortened the game clock to two minutes.
- Collisions with the star or with another player no longer end the game. They now count toward players' scores instead.
- The version that I provided to you somehow was missing the warp (a.k.a. hyperspace) functionality. I've added it to this version.

> it feels like the game tends to end really quickly, as I end up hitting either the star or the other player within about 5-10 seconds most of the time -- maybe we can remove some of the game overs and convert them to just reset the player and give score to someone? Or possibly even just go for a one minute timer, with the goal being to get the most points by then (either by destroying the other player, or reaching the goal a bunch, etc).

SpaceFight! used SpaceWar! as its starting point, including the replication of the original game play -- a ten-minute timer with the game ending on collision with the gravity well or another player. Because the screen in MakeCode Arcade is considerably smaller than the oscilloscope used on the PDP-1, it makes the MakeCode Arcade game considerably more challenging.

I retained the original gameplay to fulfill my own needs for the course that I am writing. Because this version does not need to be an actual clone, I'm happy to alter the gameplay to make it more appropriate for this platform. It's provided me with an interesting thought exercise, and it's something that I will document in my course.

> It might be worth just moving the just using on game update events to handle fuel and rockets instead of being exact about the timing; it cuts off another ~85 blocks, which can be important on computers that aren't high end: https://makecode.com/_PCueTXE5aHMm . The timings a bit off with that (e.g. sometimes you'll fire a rocket and then immediately get one back), but you can regain much of the timing either with some flags (e.g. ignore the first time it would refuel in a row flip the flag, then continue to refuel while the flag is true / set it to false when full) or by increasing the number of fuel and rocket 'units' consumed and obtained (instead of 1 every 5 seconds, use 10 at once but get them every half second, etc)

I love this idea! I did not realize that you could have multiple `game.onUpdateInterval()` routines, nor that you could use them in conjunction with `game.onUpdate()`, so that never entered my thoughts as a possibility. I've retained your modifications in this regard and built on them in this revision. The minor timing issues that you mentioned don't bother me for this game. I also will keep this in mind for future games!

> I like your 'draw text' extension a lot, I'll have to take another look at it some time

Thanks, Joey. Y'all did most of the work; I just exposed some of the functionality in Blocks form. I was worried that I was going to have to create bitmap fonts, figure out how to load them, and all of that other stuff. When I realized that y'all had already built `image.print()` routines, everything fell into place.

The GitHub repository for the extension is here:
https://github.com/robo-technical-group/pxt-arcade-display-strings

> For your extensions, you can add a variable dropdown for blocks with variables.shadow="variables_get", so you don't have holes in the blocks when users drag them out. Related: There's a minor bug preventing them from behaving perfectly - #409 - but if you omit the initial value in parens they will show up nicely for now

Excellent feedback, Joey. I've already updated my extensions with the shadows. I'll keep an eye on #409 and, when that gets corrected in production, I'll update the extensions again to suggest a more appropriate default variable.

> ~~This game really makes me think we should add a single block to set default values for controller repeat events delay and interval (related: #767). That way we can cut half the controller events in this game immediately, and also make it easy to use repeat events in a wider variety of contexts~~ (added in Microsoft/pxt-common-packages#794)

Let's revisit this again when Microsoft/pxt-common-packages#794 gets incorporated into pxt-arcade and pushed out to production. I'd like to remove the duplicate code for the controller events when that's possible, as that obviously will help with the block count.

Let me know if you'd like to make additional changes to the code to SpaceFight! and I'll be happy to incorporate them back into the pull request. Otherwise, let me know that you're happy with the game as-is, and I'll update the pull request with the current URL and code.

Thanks again for the time and energy, Joey. It means a lot.

jwunderl commented:
@alex-kulcsar There's a small error in the ``destroyPlayer2`` function (``if (true)``) but besides that it looks great to me!

![Screen Shot 2019-04-22 at 2 25 09 PM](https://user-images.githubusercontent.com/5615930/56532555-d179a780-650a-11e9-9bdc-8e1566641220.png)

If you could update the file with the new version of the code, and possibly rename it ``docs/community/spacefight.md`` (markdown file extension instead of typescript), it should be good to go for me; I can make the community page and add this to it / link the extensions in a separate pr. For now, it'll have to be displayed as a typescript game, as the docs are all saved as typescript and it seems like it has a few too many blocks to decompile for now (but the link will still show it is originally created in blocks, of course)

Also, if we could see your course when you're finished, it sounds very interesting from what I've seen on the forum / here!

(also tagging @ganicke in case I missed anything)

ganicke commented:
Change the file from *.ts to *.md and put it in the _./docs/community_ path. Edit it to also include:

```markdown
# Space Quest

` ` `blocks
...TS example source here...
` ` `
```

alex-kulcsar commented:
Good catch, Joey (@jwunderl). Given that both you and @ganicke mentioned that the file should be posted as a MarkDown file, you may want to update the instructions at https://arcade.makecode.com/game-submission, as it currently says to create a TypeScript file.

The new version of SpaceFight, which will be referenced in the updated commit, is available at https://makecode.com/_e2LM0JKsiFee

I'll update my repo shortly. Thanks again for all of your great feedback!

alex-kulcsar commented:
Since Joey was so kind to mention it here, and since I failed to do so in my previous comment (oops!), I'll briefly comment here. :-)

The course that I am writing has a ways to go yet, but I'm having a ton of fun! MakeCode Arcade has provided me with a perfect platform. The course outline is done, and I'm working on some smaller activities before diving headlong into writing.

I'm also reading Seymour Papert's _Mindstorms_ at the moment. He and Jean Piaget were such instrumental forces in the constructivist model, and I want to make sure that I am properly incorporating those ideas in my own teaching. It's been a _very_ long time since I've read material from either of them, and I figured now would be a good time to refresh my memory. :-)

I'll be sure to post about my progress in the MakeCode forums. Thanks again for the encouragement, Joey!

jwunderl commented:
Thanks again for submitting! I'll be adding an attribution at the bottom when I fix up the contribution guide and add links for this, so let me know if there's anywhere you want linked or anything else you want mentioned -- I saw http://www.alexkulcsar.com/ on your profile. (If you'd prefer to phrase it yourself, here's an example of a submitted example for micro:bit: https://makecode.microbit.org/projects/karel)

Jaqster commented:
Hi Alex! I'm the PM on the MakeCode Team, and loving that you're digging into Arcade and putting it through its paces!  MakeCode was created based on the Constructionist Learning Theory and we take a lot of our design cues from that philosophy of iterative, learning-by-doing approach.  Awesome that you are building your course incorporating this philosophy!  Keep the feedback and suggestions coming!  We are happy to continue collaborating with you and would be honored to post and help promote your work.  Thanks again for digging into Arcade!

ganicke commented:
@alex-kulcsar we're trying to make this process a little more clear: #922
